### ISSUE TITLE
File.upload() throws File I/O Error #2038 if httpstatus is not 200

### ISSUE THREAD

**Issue by MalacTheLittle** [airsdk/Adobe-Runtime-Support#2739]
We have PHP server configured to accept files with POST method under predefined id. If id doesn't exist, API return 404 status.

Now, when I try `File.upload()` and status is 404 or 401, I get Error 2044: Unhandled IOErrorEvent:. text=Error 2038: File I/O Error.
Same file uploaded with status 200 doesn't throw any error.

Tested with AIR33.1.1.929, AIR50.1.1.2, AIR50.2.3.2 on Windows

## Steps to Reproduce
File.upload() to any server that throws 404 status.

## Known Workarounds
None

**Comment by ajwfrost** [comment]
To avoid the 'unhandled IOErrorEvent' couldn't you just add an event handler for that? I'm not sure I'd see this as an error: if you make an http request (post or get) and you get a failure code (4xx etc) then IOError is the normal response?

**Comment by MalacTheLittle** [comment]
I am handling UncaughtErrorEvents in the app, but the thing is that I need to get server response with error details.

One thing I found out in the meantime is:

I have event listeners on File:
```
	file.addEventListener(DataEvent.UPLOAD_COMPLETE_DATA, apiUploadComplete);
	file.addEventListener(ProgressEvent.PROGRESS, apiUploadProgress);
	file.addEventListener(IOErrorEvent.IO_ERROR, apiUploadIOError);
	file.addEventListener(SecurityErrorEvent.SECURITY_ERROR, apiUploadSecurityError);
	file.addEventListener(HTTPStatusEvent.HTTP_RESPONSE_STATUS, apiUploadResponseStatus);
```
On upload events are triggered in next sequence:
1. apiUploadResponseStatus (HTTP_RESPONSE_STATUS)
2. apiUploadComplete (UPLOAD_COMPLETE_DATA)
and then Unhandlend error occurs.

If I change eventListener from HTTP_RESPONSE_STATUS to HTTP_STATUS, error doesn't occur and sequence of events are as follows:
1. apiUploadStatus (HTTP_STATUS)
2. apiUploadIOError (IO_ERROR)
but UPLOAD_COMPLETE_DATA event isn't triggered and I can't receive response with details from server.

Can that be considered as bug in HTTP_RESPONSE_STATUS event and can I get response in some other way while using HTTP_STATUS event?

**Comment by ajwfrost** [comment]
Ah .. yes, so as I recall, if you listen for HTTP_RESPONSE_STATUS, then the error code is signalled in that, and you should get a "complete" with the error page contents .. but yes, that would imply that you don't then end up with another IO_ERROR being sent out. Are you removing the listeners at some point? Or possibly that's happening in some internal code...

We can check on this, it presumably should work (or rather, fail/reproduce) if we just try an upload to an invalid URL within a real http server..

thanks

**Comment by MalacTheLittle** [comment]
Here's an update:
When HTTP_RESPONSE_STATUS is triggered and if I remove HTTP_RESPONSE_STATUS listener on response not 200, folowing triggered event is IOErrorEvent.IO_ERROR. Now UPLOAD_COMPLETE_DATA isn't triggered and I can't get server response.
This way Unhandled IOErrorEvent is avoided, but server response is missing.

Is there a way to get server response data through HTTP_RESPONSE_STATUS?

**Comment by ajwfrost** [comment]
So -> you get just the response code in the HTTP_RESPONSE_STATUS, and if you've registered for that event, you also get the UPLOAD_COMPLETE_DATA event. If there's an error condition, then you also will get the IO_ERROR event.

The documentation seems to suggest this is the expected order.. at least for URLStream (see also #129):
> In AIR, a URLRequest object can register for the httpResponse status event. Unlike the httpStatus event, the httpResponseStatus event is delivered before any response data. Also, the httpResponseStatus event includes values for the responseHeaders and responseURL properties (which are undefined for an httpStatus event. Note that the httpResponseStatus event (if any) will be sent before **(and in addition to)** any complete or error event.

But the documentation for the upload event isn't very clear on this:
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/net/FileReference.html#upload()

It seems to suggest only a limited number of conditions when you'd expect an 'ioError' event. So .. we could easily change this so that if you've got an HTTP_RESPONSE_STATUS handler, we send the UPLOAD_COMPLETE_DATA event and don't send the IO_ERROR one. But that might impact people who are already handling this scenario.

Probably the best option is to listen for HTTP_RESPONSE_STATUS handler, and if the http status code is an error one, wait until you get the IO_ERROR event handler before you start removing the event listeners and closing down the objects. If it's a success code, you can do that in the UPLOAD_COMPLETE_DATA..

**Comment by MalacTheLittle** [comment]
> So -> you get just the response code in the HTTP_RESPONSE_STATUS, and if you've registered for that event, you also get the UPLOAD_COMPLETE_DATA event. If there's an error condition, then you also will get the IO_ERROR event.

Right, just tested if I don't register UPLOAD_COMPLETE_DATA, but do register IO_ERROR, Unhandled IOErrorEvent won't trigger.

> Probably the best option is to listen for HTTP_RESPONSE_STATUS handler, and if the http status code is an error one, wait until you get the IO_ERROR event handler before you start removing the event listeners and closing down the objects. If it's a success code, you can do that in the UPLOAD_COMPLETE_DATA..

This is how I set temporary solution, but I don't have response message from server since IO_ERROR doesn't provide it. Therefore, for solution I suggest to add responseBody to HTTP_RESPONSE_STATUS as a new property. In this way, there would be no impact on current handling scenarios. 
And as I can see from Postman, it's not an unusual practice:

![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/56675407/d179f4df-7d42-4d08-89f7-61eec7205bb9)

**Comment by ajwfrost** [comment]
For this:
> Right, just tested if I don't register UPLOAD_COMPLETE_DATA, but do register IO_ERROR, Unhandled IOErrorEvent won't trigger.

You should register them both .. then you will get the response data from the first event, and the error comes in the second event. When we trigger the http status event, we don't necessarily have the full response data..

**Comment by MalacTheLittle** [comment]
Ahhh, OK! Now I see what I was doing wrong. 
On UPLOAD_COMPLETE_DATA I removed all listeners (IO_ERROR also) and that's why I got Unhandled IOErrorEvent. I wasn't aware of event sequence: 1. HTTP_RESPONSE_STATUS, 2. UPLOAD_COMPLETE_DATA, 3. IO_ERROR. 
I find this sequence a bit confusing since I would expect that UPLOAD_COMPLETE_DATA shouldn't happen if IO_ERROR occurs, but in order to get response data, it has it's logic.

Thanks Andrew!
